### PR TITLE
General power operator x ** y #468

### PR DESCRIPTION
**Pull request by archibate**

### CODE REVIEW

**Comment by archibate** [comment]
@yuanming-hu Thank you for your approval! This is the first github project I contribute with, my commits were approved and merged, being invited the dev team... I couldn't appreciate it more.

**Comment by archibate** [comment]
BTW, I see
https://github.com/taichi-dev/taichi/blob/d76c64fc27d32daeb4446f204c2b8000877dea3a/python/taichi/lang/expr.py#L245
https://github.com/taichi-dev/taichi/blob/d76c64fc27d32daeb4446f204c2b8000877dea3a/python/taichi/lang/ops.py#L39
These two functions are doing exactly the same thing!

We can do add by `a + b`, so `ti.add(a, b)` is not necessary.
We can do mod by `a % b`, so `ti.mod(a, b)` is not necessary, but C-style `ti.raw_mod(a, b)` one is.
We can do pow by `a ** b`, so what's the point of `ti.pow(a, b)`?

**Comment by yuanming-hu** [comment]
That makes sense! Let's remove `ti.pow` then. Thanks so much for your lightning-fast implementation! Taking a look now.

(It's also my first time organizing a community of contributors. I really appreciate contributions from you guys. I'm learning/trying a lot of new skills to make sure every contributor is happy coding :-) Glad that you are having fun!)

File: tests/python/test_pow.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  
+  for x in [2, 1.4, 3, 5.11, 6.12, 10.0]:
+    for y in [2, 0.5, 4, 8, 1.12, -1, -9, 2.01, -3.3]:
+      print(x, '**', y)
+      func(x, y)
+      print(z[None], 'vs', x ** y)
+      assert abs(z[None] / x ** y - 1) < 0.0001
+      assert abs(w[None] / pow(x, y) - 1) < 0.0001
+
+test_pow_scan()
```
Comment (yuanming-hu) [inline tests/python/test_pow.py]: `pytest` (invoked by `ti test_python`) will automatically collect all functions that start with `test_`, so this line is not really necessary :)

File: tests/python/test_pow.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import taichi as ti
+
+
+@ti.all_archs
+def test_pow_scan():
+  z = ti.var(ti.f32, shape=())
```
Comment (yuanming-hu) [inline tests/python/test_pow.py]: Here we can probably test the operator with precisions `ti.f64`, `ti.i32`, `ti.i64` as well.

**Comment by yuanming-hu** [comment]
How about this: for integer pow, let's add a function to the runtime `pow_i32/i64(x, y)` using [exponential by squaring](https://stackoverflow.com/questions/101439/the-most-efficient-way-to-implement-an-integer-based-power-function-powint-int) so that we get 
 - accuracy. No float point error issues anymore
 - speed. `exponential by squaring` has `O(log y)` performance instead of `O(y)`

**Comment by archibate** [comment]
Good idea! But we may want to `pow(float, int)` instead of `pow(int, int)`.
How about name them like `powi32_f32(f32, i32)` and `powi32_f64(f64, i32)`?

**Comment by archibate** [comment]
A lot of work to do but, I must go to sleep now (in my time zone). See you tomorrow!

**Comment by yuanming-hu** [comment]
Good night! Thanks a lot for the hard work!!

I think we only need to support `i32 ** i32, i64 ** i64, f32 ** f32, f64 ** f64`. For `f32 ** i32`, the type checker will promote it into `f32 ** f32`.

**Comment by archibate** [comment]
Good morning! Guess what? Oh, I can't even fall asleep yesternight! Both literally or physically!

Yeah but I personally think `f32 ** i32` is more important than `i32 ** i32` since, eg. we usually need to compute `x ** 2` or `x ** 3`, where `x` is float or integer, and the second argument is often integer.
If user want to do `x ** 233`, then a optimization technique is really necessary. So we should support `float ** int`, for  _exponential by squaring_.

**Comment by yuanming-hu** [comment]
I usually suffer from bad sleep quality as well, if I go to bed with a problem to think about/exiciting news to recap :-)

Your concerns are very valid. I feel like for cases like `x ** 2` or `x ** 3`, the exponents are usually compile-time constant, which will directly cause the `pow` opeartor to get unrolled at compile-time: https://github.com/taichi-dev/taichi/blob/d76c64fc27d32daeb4446f204c2b8000877dea3a/python/taichi/lang/expr.py#L250

(Btw, we'd better add a threshold here, otherwise if somebody does `x ** 10000000000000` Taichi will generate too many instructions and explode.)

**Comment by yuanming-hu** [comment]
(In fact, I believe `std::pow(f32, f32)` already optimizes for the cases with `2`, `3`, `0.5` as exponents. For `x ** 233`, `std::pow(f32, f32)` is likely doing a better job than exponentaion by squaring.) That's I think why we need to deal with `pow(i32, i32)` only.

**Comment by yuanming-hu** [comment]
For example, the `glibc` implementation: https://github.com/lattera/glibc/blob/895ef79e04a953cac1493863bcae29ad85657ee1/sysdeps/ieee754/dbl-64/e_pow.c#L75

**Comment by archibate** [comment]
That might be true. I wrote a test code:
```cpp
#include <iostream>
#include <cmath>
#include <ctime>

int main(void)
{
  clock_t c, d;
#define TEST(x, y) \
  c = clock(); \
  for (int i = 0; i < 10000000; i++) { \
    std::pow(x, y); \
    std::pow(x, y); \
    std::pow(x, y); \
  } \
  d = c; \
  c = clock(); \
  std::cout << #x " ** " #y " = " << std::pow(x, y); \
  std::cout << " : " << float(c - d) / CLOCKS_PER_SEC << std::endl;

  TEST(2, 3)
  TEST(2.0, 3)
  TEST(2, 3.0)
  TEST(2.0, 3.0)
  TEST(2.01, 3)
  TEST(2, 3.01)
  TEST(2.01, 3.01)
}
```
And the result is:
```
2 ** 3 = 8 : 1.01514
2.0 ** 3 = 8 : 0.628173
2 ** 3.0 = 8 : 0.636521
2.0 ** 3.0 = 8 : 0.019933
2.01 ** 3 = 8.1206 : 0.633067
2 ** 3.01 = 8.05564 : 0.636888
2.01 ** 3.01 = 8.17749 : 0.020306
```

**Comment by yuanming-hu** [comment]
Cool! Are you compiling with optimization on? If so the compiler might just optimize away all the `pow`'s since their results are not used. I would compile with optimizations (-O3) allocate a local variable to store the sum of all the `pow`'s, and finally print the sum out to make sure the compiler is not optimizing out the computation.

**Comment by yuanming-hu** [comment]
Also `clock` might not be the most accurate function to get time. See the accurate timer implementation for Unix/Windows: https://github.com/taichi-dev/taichi/blob/master/taichi/system/timer.cpp#L25

**Comment by archibate** [comment]
Of course, I did specified `-O0`.

```cpp
#include <iostream>
#include <cmath>
#include <sys/time.h>

namespace quick
{
  template <typename T>
  T pow(T x, int n)
  {
    T tmp = x;
    T ans = 1;
    while (n) {
      if (n & 1) ans *= tmp;
      tmp = tmp * tmp;
      n >>= 1;
    }
    return ans;
  }
};

namespace slow
{
  template <typename T>
  T pow(T x, int n)
  {
    T ans = 1;
    while (n--) {
      ans *= x;
    }
    return ans;
  }
};

double get_time() {
  struct timeval tv;
  gettimeofday(&tv, nullptr);
  return tv.tv_sec + 1e-6 * tv.tv_usec;
}

int main(void)
{
  double c, d;
#define TEST(z, x, y) \
  c = get_time(); \
  for (int i = 0; i < 10000000; i++) { \
    z::pow(x, y); \
    z::pow(x, y); \
    z::pow(x, y); \
  } \
  d = c; \
  c = get_time(); \
  std::cout << #z "::pow(" #x ", " #y ") = " << std::pow(x, y); \
  std::cout << " : " << c - d << std::endl;

  TEST(quick, 2, 9)
  TEST(quick, 2.0, 9)
  TEST(quick, 2.01, 9)
  TEST(slow, 2, 9)
  TEST(slow, 2.0, 9)
  TEST(slow, 2.01, 9)
  TEST(std, 2, 9)
  TEST(std, 2.0, 9)
  TEST(std, 2, 9.0)
  TEST(std, 2.0, 9.0)
  TEST(std, 2.01, 9)
  TEST(std, 2, 9.01)
  TEST(std, 2.01, 9.01)
}
```

Result:
```
quick::pow(2, 9) = 512 : 0.268731
quick::pow(2.0, 9) = 512 : 0.295024
quick::pow(2.01, 9) = 535.506 : 0.294242
slow::pow(2, 9) = 512 : 0.434631
slow::pow(2.0, 9) = 512 : 0.454374
slow::pow(2.01, 9) = 535.506 : 0.453083
std::pow(2, 9) = 512 : 1.01418
std::pow(2.0, 9) = 512 : 0.631115
std::pow(2, 9.0) = 512 : 0.63939
std::pow(2.0, 9.0) = 512 : 0.0201051
std::pow(2.01, 9) = 535.506 : 0.634391
std::pow(2, 9.01) = 515.561 : 0.643357
std::pow(2.01, 9.01) = 539.258 : 0.0203102
```

Seems our `quick::pow(f, i)` is far more good than `std::pow(f, f)`.

**Comment by yuanming-hu** [comment]
Actually, I would benchmark with `-O3`, since that's the real performance when we invoke that function.

**Comment by yuanming-hu** [comment]
...but when you compile this piece of code with `-O3`, the optimizer will probably directly replace `std::pow(2.0, 9.0)` with `512.0`... Measuring performance is hard.

I would do something like
```
int global=0;

...

#define TEST(x, y) \
double s0=0;\
for (int i = 0; i < 10000000; i++) {\
    s += z::pow(x, y + int(1e-20 * i + global)); // stop the compiler from optizing pow out \
  }\
printf("%\f\n", s);
```

**Comment by archibate** [comment]
How about using `rand()`?

**Comment by yuanming-hu** [comment]
> How about using `rand()`?

That's a good idea for benchmarking pow with general inputs, but here you are studying its performance with fixed inputs.

**Comment by archibate** [comment]
```cpp
#include <cmath>
#include <iostream>
#include <sys/time.h>

namespace quick {
        template <typename T>
                T pow(T x, int n)
                {
                        T tmp = x;
                        T ans = 1;
                        while (n) {
                                if (n & 1) ans *= tmp;
                                tmp = tmp * tmp;
                                n >>= 1;
                        }
                        return ans;
                }
};
namespace slow {
        template <typename T>
                T pow(T x, int n)
                {
                        T ans = 1;
                        while (n--) {
                                ans *= x;
                        }
                        return ans;
                }
};
namespace ___ {
        template <typename T>
                T pow(T x, int n)
                {
                        return x * n;
                }
};

double get_time()
{
  struct timeval tv;
  gettimeofday(&tv, nullptr);
  return tv.tv_sec + 1e-6 * tv.tv_usec;
}

int main(void)
{
        double c, d;
        float s = 0;

#define TEST(z) \
        c = get_time(); \
        for (int i = 0; i < 100000; i++) { \
                int n = rand(); \
                float x = n * 2 / RAND_MAX; \
                int y = n % 8; \
                s += z::pow(x, y); \
        } \
        d = c; \
        c = get_time(); \
        std::cout << #z "\t" << c - d << std::endl; \
        std::cout << s << "\r\033[K"; \

        TEST(___)
        TEST(std)
        TEST(quick)
        TEST(slow)

        return 0;
}
```
-O0:
```
___     0.00122595
std     0.00224185
quick   0.00341392
slow    0.00334191
```
-O3:
```
___     0.000956059
std     0.001791
quick   0.00248003
slow    0.00236607
```

**Comment by archibate** [comment]
果然是之前那个测试代码的问题。

**Comment by archibate** [comment]
我想`std::pow`作为万人用的标准库，应该还是会针对不同情况优化的，而且我估计`std::pow<int, int>`和`std::pow<float, int>`是存在的。

**Comment by archibate** [comment]
But how we do with cuda? I followed `__nv_atan2f` and guessed `__nv_powf`. Is there a corr `__nv_powi`?

**Comment by yuanming-hu** [comment]
I believe we can trust the implementation of standard `pow` implementations, on both CPUs and GPUs :-) There's indeed a powi: https://docs.nvidia.com/cuda/libdevice-users-guide/__nv_powi.html#__nv_powi

**Comment by yuanming-hu** [comment]
There's indeed a `pow(f32, i32)` in std: https://en.cppreference.com/w/cpp/numeric/math/pow

**Comment by archibate** [comment]
> `double @__nv_powi(double %x, i32 %y)`
But the first argument is double. So it can't be directly called from (i, i) context.

**Comment by archibate** [comment]
> Your concerns are very valid. I feel like for cases like x ** 2 or x ** 3, the exponents are usually compile-time constant, which will directly cause the pow opeartor to get unrolled at compile-time.

By this we are not very concerned with `f ** i` performance, so `__nv_powi` may be simply ignored.

**Comment by yuanming-hu** [comment]
> By this we are not very concerned with f ** i performance, so __nv_powi may be simply ignored.

Exactly! I personally don't want to get things too complicated. Let's just trust `std::pow(f32, f32  /  f64, f64)`.

File: taichi/runtime/runtime.cpp [hunk 3]
```diff
@@ -207,6 +215,16 @@ f64 pow_f64(f64 a, f64 b) {
   return std::pow(a, b);
 }
 
+#if ARCH_cuda
```
Comment (yuanming-hu) [inline taichi/runtime/runtime.cpp]: For `pow_i32` and `nv_pow_i32` we may want to use exponentiation by squaring, for accuracy reasons. `std::pow` might nit have enough accuracy here.

Comment (yuanming-hu) [reply taichi/runtime/runtime.cpp]: ...and then `nv_pow_x` and `pow_x` can get merged since they are the same.

**Comment by archibate** [comment]
BTW, will _exponential by squaring_ in Expr.__pow__ help? Will it be unrolled successfully? Will it gain perf benefits?

**Comment by yuanming-hu** [comment]
> BTW, will exponential by squaring in Expr.pow help? Will it be unrolled successfully? Will it gain perf benefits?

It will get unrolled. I think that would be great if you implement that as well.

I would also change the range `[0, 12)` to `[-6, 6]`, since something people may do `x ** -2`.

**Comment by archibate** [comment]
Sometimes people wants more than 12. eg. Doing light tracing for phong material, `pow(dot(r, n), shininess)`, where `shininess` could be very big, up to 50.

File: taichi/backends/codegen_llvm.h [hunk 4]
```diff
@@ -495,6 +495,44 @@ class CodeGenLLVM : public IRVisitor, public ModuleBuilder {
...
+      } else if (current_arch() == Arch::cuda) {
+        if (ret_type == DataType::f32) {
+          stmt->value =
+              create_call("__nv_powf", {stmt->lhs->value, stmt->rhs->value});
+        } else if (ret_type == DataType::f64) {
+          stmt->value =
+              create_call("__nv_pow", {stmt->lhs->value, stmt->rhs->value});
+        } else if (ret_type == DataType::i32) {
+          stmt->value =
+              create_call("nv_pow_i32", {stmt->lhs->value, stmt->rhs->value});
```
Comment (yuanming-hu) [inline taichi/backends/codegen_llvm.h]: Looks good! but should here be `pow_i32` instead of `nv_pow_i32`?

Comment (archibate) [reply taichi/backends/codegen_llvm.h]: Yeah, that's it.

**Comment by yuanming-hu** [comment]
Let's test the cases with compile-time constant exponents as well. E.g. 

```
@ti.kernel
  def func3(x: dt):
    z[None] = x ** 3
```

Or if you use Taichi template meta-programming
```
@ti.kernel
  def func(x: dt, y: ti.template()):
    z[None] = x ** y # here y is compile-time constant

func(3, 1)
func(3, 2)
func(3, 3)
func(3, 4)
func(3, -4)
```

> `shininess` could be very big, up to 50.

Oh that's very true. I almost forgot this use case. I'm not sure if exponential by squaring is still faster than `std::pow` for exponent 50 though.

**Comment by archibate** [comment]
> `ti.template()`

Cool!

**Comment by archibate** [comment]
Python exp-by-sqr run into a segfault.

**Comment by yuanming-hu** [comment]
Let me take a look

**Comment by archibate** [comment]
```
[E 02/16/20 08:27:30.807] [python_bindings.cpp:expr_assign@27] lhs->is_lvalue()
[E 02/16/20 08:27:30.807] Received signal 6 (Aborted)
***********************************
* Taichi Compiler Stack Traceback *                                                                         
***********************************                                                                         
/home/bate/Develop/taichi/build/taichi_core.so: taichi::signal_handler(int)                                 
/usr/lib/libc.so.6(+0x3bfb0) [0x7fcfa5320fb0]                                                               
/usr/lib/libc.so.6: gsignal                                                                                 
/home/bate/Develop/taichi/build/taichi_core.so: taichi::Tlang::expr_assign(taichi::Tlang::Expr const&, taichi::Tlang::Expr const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)     
/home/bate/Develop/taichi/build/taichi_core.so(+0x7b58dc) [0x7fcf9599f8dc]                                  
/home/bate/Develop/taichi/build/taichi_core.so(+0x7b57ff) [0x7fcf9599f7ff]                                  
/home/bate/Develop/taichi/build/taichi_core.so(+0x643347) [0x7fcf9582d347]                                  
/usr/lib/libpython3.8.so.1.0: PyCFunction_Call                                                              
/usr/lib/libpython3.8.so.1.0: _PyObject_MakeTpCall                                                          
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault                                                      
/usr/lib/libpython3.8.so.1.0: _PyFunction_Vectorcall                                                        
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault                                                      
/usr/lib/libpython3.8.so.1.0: _PyFunction_Vectorcall                                                        
/usr/lib/libpython3.8.so.1.0: _PyObject_FastCall_Prepend                                                    
/usr/lib/libpython3.8.so.1.0(+0x15d843) [0x7fcfa50ca843]                                                    
/usr/lib/libpython3.8.so.1.0(+0xb5517) [0x7fcfa5022517]                                                     
/usr/lib/libpython3.8.so.1.0(+0x1b1e67) [0x7fcfa511ee67]                                                    
/usr/lib/libpython3.8.so.1.0: PyNumber_InPlaceMultiply                                                      
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault                                                      
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName                                                      
/usr/lib/libpython3.8.so.1.0: _PyFunction_Vectorcall                                                        
/usr/lib/libpython3.8.so.1.0: _PyObject_FastCall_Prepend                                                    
/usr/lib/libpython3.8.so.1.0(+0x15dd47) [0x7fcfa50cad47]                                                    
/usr/lib/libpython3.8.so.1.0(+0x1e46d3) [0x7fcfa51516d3]                                                    
/usr/lib/libpython3.8.so.1.0: PyNumber_Power                                                                
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault                                                      
/usr/lib/libpython3.8.so.1.0: _PyFunction_Vectorcall                                                        
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault                                                      
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName                                                      
/usr/lib/libpython3.8.so.1.0: _PyFunction_Vectorcall                                                        
/usr/lib/libpython3.8.so.1.0: PyVectorcall_Call                                                             
/home/bate/Develop/taichi/build/taichi_core.so(+0x7afc25) [0x7fcf95999c25]                                  
/home/bate/Develop/taichi/build/taichi_core.so(+0x7afbbf) [0x7fcf95999bbf]                                  
/home/bate/Develop/taichi/build/taichi_core.so: taichi::Tlang::Kernel::Kernel(taichi::Tlang::Program&, std::function<void ()>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool)   
/home/bate/Develop/taichi/build/taichi_core.so: std::_MakeUniq<taichi::Tlang::Kernel>::__single_object std::make_unique<taichi::Tlang::Kernel, taichi::Tlang::Program&, std::function<void ()> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool&>(taichi::Tlang::Program&, std::function<void ()> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool&)                                                                                                
/home/bate/Develop/taichi/build/taichi_core.so(+0x74f7bc) [0x7fcf959397bc]                                  
/home/bate/Develop/taichi/build/taichi_core.so(+0x643347) [0x7fcf9582d347]                                  
/usr/lib/libpython3.8.so.1.0: PyCFunction_Call                                                              
/usr/lib/libpython3.8.so.1.0: _PyObject_MakeTpCall                                                          
/usr/lib/libpython3.8.so.1.0(+0xbd294) [0x7fcfa502a294]                                                     
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault                                                      
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName                                                      
/usr/lib/libpython3.8.so.1.0(+0x16220d) [0x7fcfa50cf20d]                                                    
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault                                                      
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName                                                      
/usr/lib/libpython3.8.so.1.0: _PyObject_FastCallDict                                                        
/usr/lib/libpython3.8.so.1.0: _PyObject_Call_Prepend                                                        
/usr/lib/libpython3.8.so.1.0(+0x1ed87a) [0x7fcfa515a87a]                                                    
/usr/lib/libpython3.8.so.1.0: _PyObject_MakeTpCall                                                          
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault                                                      
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName                                                      
/usr/lib/libpython3.8.so.1.0: PyEval_EvalCodeEx                                                             
/usr/lib/libpython3.8.so.1.0: PyEval_EvalCode                                                               
/usr/lib/libpython3.8.so.1.0(+0x1ed7ea) [0x7fcfa515a7ea]                                                    
/usr/lib/libpython3.8.so.1.0(+0x229325) [0x7fcfa5196325]                                                    
/usr/lib/libpython3.8.so.1.0: PyRun_FileExFlags                                                             
/usr/lib/libpython3.8.so.1.0: PyRun_SimpleFileExFlags                                                       
/usr/lib/libpython3.8.so.1.0: Py_RunMain                                                                    
/usr/lib/libpython3.8.so.1.0: Py_BytesMain                                                                  
/usr/lib/libc.so.6: __libc_start_main                                                                       
python(_start+0x2e) [0x56549327e05e]
```

File: python/taichi/lang/expr.py [hunk 5]
```diff
@@ -243,13 +243,22 @@ def atomic_add(self, other):
...
-    ret = self
-    for i in range(power - 1):
-      ret = ret * self
-    return ret
+    negative = power < 0
+    tmp = self
+    ret = 1
+    while power:
+      if power & 1:
+        ret *= tmp
```
Comment (yuanming-hu) [inline python/taichi/lang/expr.py]: Could you try change this to `ret = ti.expr_init(ret, tmp)`?

File: python/taichi/lang/expr.py [hunk 6]
```diff
@@ -243,13 +243,22 @@ def atomic_add(self, other):
...
+      tmp *= tmp
```
Comment (yuanming-hu) [inline python/taichi/lang/expr.py]: and this line to `tmp = ti.expr_init(tmp * tmp)`?

**Comment by archibate** [comment]
`TypeError: expr_init() takes 1 positional argument but 2 were given`

File: python/taichi/lang/expr.py (hunk 6 above)
Comment (archibate) [reply python/taichi/lang/expr.py]: It works!

**Comment by yuanming-hu** [comment]
Sorry, should be `ret = ti.expr_init(ret * tmp)`

**Comment by archibate** [comment]
```
$ python pow.my.py
[Taichi version 0.5.0, cpu only, commit 52bb0020]
3.0
9.0
27.0
81.0
0.012345679104328156
```
!!!!

**Comment by archibate** [comment]
What did `ti.expr_init` do?

**Comment by yuanming-hu** [comment]
Long story short: it will try to create a (assignable) local variable and store your results there.

**Comment by archibate** [comment]
> Long story short: it will try to create a (assignable) local variable and store your results there.

That's why `lhs->is_lvalue()` returns false, lhs is rvalue!

**Comment by archibate** [comment]
So `tmp = ti.expr_init(tmp * tmp)` actually not necessary? Since only the last `ret` need to be lvalue.

**Comment by archibate** [comment]
```python
    ret = 1
    while power:
      if power & 1:
        ret = ret * tmp
      tmp = tmp * tmp
      power >>= 1
    ret = ti.expr_init(ret)
```
works.

**Comment by yuanming-hu** [comment]
Yeah it might work but if you `ti.init(print_ir=True)`, you might find the benefit of exponentiation by squaring is lost. It's essential to store the local results in a local variable.

**Comment by archibate** [comment]
Seems no difference.

Without `ti.expr_init`:
```
==========
kernel {
  $0 = offloaded  {
    <f32x1>  $1 = arg[0]
    <i32x1>  $2 = const [1]
    <f32x1>  $3 = cast<float32> $2
    <f32x1>  $4 = mul $1 $3
    <f32x1>  $5 = mul $1 $1
    <f32x1>  $6 = mul $5 $5
    <f32x1>  $7 = mul $6 $6
    <f32x1>  $8 = mul $4 $7
    $9 = get root
    <i32x1>  $10 = linearized(ind {}, stride {})
    $11 = [S0root][root]::lookup($9, $10) coord = {} activate = false
    $12 = get child [S0root->S1dense] $11
    $13 = [S1dense][dense]::lookup($12, $10) coord = {} activate = false
    $14 = get child [S1dense->S2place] $13
    <f32x1>  $15 : global store [$14 <- $8]
  }
}
==========
```

With `ti.expr_init`:
```
==========
kernel {
  $0 = offloaded  {
    <f32x1>  $1 = arg[0]
    <i32x1>  $2 = const [1]
    <f32x1>  $3 = cast<float32> $2
    <f32x1>  $4 = mul $1 $3
    <f32x1>  $5 = mul $1 $1
    <f32x1>  $6 = mul $5 $5
    <f32x1>  $7 = mul $6 $6
    <f32x1>  $8 = mul $4 $7
    $9 = get root
    <i32x1>  $10 = linearized(ind {}, stride {})
    $11 = [S0root][root]::lookup($9, $10) coord = {} activate = false
    $12 = get child [S0root->S1dense] $11
    $13 = [S1dense][dense]::lookup($12, $10) coord = {} activate = false
    $14 = get child [S1dense->S2place] $13
    <f32x1>  $15 : global store [$14 <- $8]
  }
}
==========
```

**Comment by archibate** [comment]
Seems `!lhs->is_value()` problem comes from `*=`.

**Comment by archibate** [comment]
```
    <i32x1>  $2 = const [1]
    <f32x1>  $3 = cast<float32> $2
    <f32x1>  $4 = mul $1 $3
```
Too bad, may be solved with #456.

**Comment by yuanming-hu** [comment]
Are these IR before optimization passes or after? The simplification passes might do the squaring optimization as well.

Yeah, `*=` has an issue if the LHS is not a lvalue.

**Comment by archibate** [comment]
> Are these IR before optimization passes or after? The simplification passes might do the squaring optimization as well.

Yeah, it's after. I scrolled up and see code before optimization:
```
==========
kernel {
  <f32x1>  $0 = arg[0]
  <i32x1>  $1 = const [1]
  <f32x1>  $2 = cast<float32> $1
  <f32x1>  $3 = mul $0 $2
  <f32x1>  $4 = arg[0]
  <f32x1>  $5 = arg[0]
  <f32x1>  $6 = mul $5 $5
  <f32x1>  $7 = arg[0]
  <f32x1>  $8 = arg[0]
  <f32x1>  $9 = mul $8 $8
  <f32x1>  $10 = mul $9 $9
  <f32x1>  $11 = arg[0]
  <f32x1>  $12 = arg[0]
  <f32x1>  $13 = mul $12 $12
  <f32x1>  $14 = arg[0]
  <f32x1>  $15 = arg[0]
  <f32x1>  $16 = mul $15 $15
  <f32x1>  $17 = mul $16 $16
  <f32x1>  $18 = mul $17 $17
  <f32x1>  $19 = mul $3 $18
  <f32x1>  $20 = global ptr [S2place], index [] activate=true
  <f32x1>  $21 : global store [$20 <- $19]
}
==========
```

**Comment by archibate** [comment]
Aborted when test_pow_i64

**Comment by yuanming-hu** [comment]
It seems that the global store instruction is failing. Could you share with me the optimized IR before codegen?

**Comment by archibate** [comment]
`ti.init(print_ir=True)` not work. Where should I add?

**Comment by yuanming-hu** [comment]
At the very beginning

**Comment by archibate** [comment]
can't find, even codegen_llvm::visit not called.

**Comment by archibate** [comment]
Found, in `CodeGenLLVM::emit_to_module`, is reachable. What code should be add?

**Comment by yuanming-hu** [comment]
I have to go AFK but it's worth checking the IR to see if an i64 result is stored to an i32 pointer.

**Comment by archibate** [comment]
Ok, `irpass::print(kernel->ir)`:
```
==========
kernel {
  $0 = offloaded  {
    <i32x1>  $1 = const [1]
    <i64x1>  $2 = cast<int64> $1
    $3 = get root
    <i32x1>  $4 = linearized(ind {}, stride {})
    $5 = [S0root][root]::lookup($3, $4) coord = {} activate = false
    $6 = get child [S0root->S1dense] $5
    $7 = [S1dense][dense]::lookup($6, $4) coord = {} activate = false
    $8 = get child [S1dense->S2place] $7
    <i64x1>  $9 : global store [$8 <- $2]
  }
}
==========
```

**Comment by archibate** [comment]
pow_i64 too hard to fix. maybe later, time for breakfast.

**Comment by yuanming-hu** [comment]
I'm taking a look at this. Seems an internal Taichi issue.

**Comment by yuanming-hu** [comment]
Fixed. Let's see if all tests pass now.

**Pull request merged**